fontbe/src/cpal.rs

```diff
@@ -0,0 +1,87 @@
+//! Generates a [cpal](https://learn.microsoft.com/en-us/typography/opentype/spec/cpal) table.
+
+use crate::{
+    error::Error,
+    orchestration::{AnyWorkId, BeWork, Context, WorkId},
+};
+use fontdrasil::orchestration::{Access, Work};
+use fontir::orchestration::WorkId as FeWorkId;
+use log::debug;
+use write_fonts::{tables::cpal::ColorRecord, tables::cpal::Cpal, NullableOffsetMarker};
+
+#[derive(Debug)]
+struct CpalWork {}
+
+pub fn create_cpal_work() -> Box<BeWork> {
+    Box::new(CpalWork {})
+}
+
+impl Work<Context, AnyWorkId, Error> for CpalWork {
+    fn id(&self) -> AnyWorkId {
+        WorkId::Cpal.into()
+    }
+
+    fn read_access(&self) -> Access<AnyWorkId> {
+        Access::Variant(AnyWorkId::Fe(FeWorkId::ColorPalettes))
+    }
+
+    /// Generate [cpal](https://learn.microsoft.com/en-us/typography/opentype/spec/cpal)
+    fn exec(&self, context: &Context) -> Result<(), Error> {
+        let colors = context.ir.colors.try_get();
+        // Guard clause: no colors
+        if !colors
+            .as_ref()
+            .map(|c| c.palettes.iter().any(|p| !p.is_empty()))
+            .unwrap_or_default()
+        {
+            debug!("Skip cpal; no colors");
+            return Ok(());
+        }
+        let colors = colors.unwrap();
```

Comment:
I was staring at this for a while trying to figure out how I'd write it and I think I prefer,

```suggestion
        let colors = match context.ir.colors.try_get() {
            Some(c) if !c.palettes.iter().all(Vec::is_empty) => c,
            _ => {
                debug!("Skip cpal; no colors");
                return Ok(());
            }
        };
```

But this makes me wonder why we are even getting here in the case where the palettes are all empty? Why don't we validate that at construction time, and just not bother setting colors at all in that case?

Reply:
Agreed, moved check